Realm.Shared/linq/RealmResultsVisitor.cs

```diff
@@ -179,44 +216,60 @@ internal override Expression VisitMethodCall(MethodCallExpression m)
+                        return MakeDefaultWithTransactionIfNeeded(_metadata);
+                    }
                     var firstRow = Realm.CreateRowHandle(firstRowPtr, _realm.SharedRealmHandle);
                     IntPtr nextIndex = (IntPtr)(firstRow.RowIndex+1);
                     var nextRowPtr = _coreQueryHandle.FindDirect(nextIndex);
                     if (nextRowPtr != IntPtr.Zero)
                         throw new InvalidOperationException("Sequence contains more than one matching element");
                     return Expression.Constant(_realm.MakeObjectForRow(_metadata, firstRow));
                 }
-                if (m.Method.Name == nameof(Queryable.ElementAt))
+                if (m.Method.Name.StartsWith("Last"))
                 {
-                    Visit(m.Arguments.First());
-                    var index = (int)ExtractConstantValue(m.Arguments.Last());
+                    RecurseToWhereOrRunLambda(m); 
 
-                    RowHandle row = null;
-                    if (OptionalSortDescriptorBuilder == null)
+                    var lastRowPtr = IntPtr.Zero;
+                    using (ResultsHandle rh = _realm.MakeResultsForQuery(_coreQueryHandle, OptionalSortDescriptorBuilder)) 
                     {
-                        var rowPtr = _coreQueryHandle.FindDirect((IntPtr)index);
-                        row = Realm.CreateRowHandle(rowPtr, _realm.SharedRealmHandle);
-                    }
-                    else
-                    {
-                        using (ResultsHandle rh = _realm.MakeResultsForQuery(_coreQueryHandle, OptionalSortDescriptorBuilder))
-                        {
-                            var rowPtr = rh.GetRow(index);
-                            row = Realm.CreateRowHandle(rowPtr, _realm.SharedRealmHandle);
-                        }
+                        var lastIndex = rh.Count() - 1;
+                        if (lastIndex >= 0)
+                            lastRowPtr = rh.GetRow(lastIndex);
                     }
+                    if (lastRowPtr != IntPtr.Zero)
+                        return Expression.Constant(_realm.MakeObjectForRow(_metadata, lastRowPtr));
+                    if (m.Method.Name == "Last")
+                        throw new InvalidOperationException("Sequence contains no matching element");
+                    Debug.Assert (m.Method.Name == "LastorDefault");
+                    return MakeDefaultWithTransactionIfNeeded(_metadata);
+                }
+                if (m.Method.Name == nameof(Queryable.ElementAt))
+                {
+                    var row = VisitElementAt(m);
+                    if (row == null || row.IsInvalid)
+                        throw new ArgumentOutOfRangeException();
+                    return Expression.Constant(_realm.MakeObjectForRow(_metadata, row));
+                }
+                if (m.Method.Name == nameof(Queryable.ElementAtOrDefault))
```

Comment:
Why is this different from the rest? It seems it's a matter of an if-case inside the `if (row == null || row.IsInvalid)`